### DISCUSSION TITLE
Obtaining runoff results from all subcatchments in "for step in sim"

### DISCUSSION THREAD

**Discussion by vww-wrc** [pyswmm/pyswmm#322] [Q&A] [answered]
**Describe the bug**
I am running into problems of getting pyswmm to produce the result of all subcatchments in a simulation run by stepping through timesteps. Basically I am trying to loop through a list of new LID parameter values to explore how subcatchment runoff changes due to these LID new values. I was able to set the new LID values, but failing to 1) obtain results of all subcatchments' runoff in a single simulation, and 2) running loops of simulation with results saving to files with different names. 

**Code:**
sim = Simulation("template.inp")  ## the SWMM model has 100 subcatchments S1-S100.

####create subcatchment ID list 
subLst = []
for sub in Subcatchments(sim):
    # print(sub.subcatchmentid)
    subLst.append(sub.subcatchmentid)
print(subLst)

####initialise pyswmm.subcatchments.Subcatchments
subcat = Subcatchments(sim)

####loop through subcatchment list to run simulation 
for s in subLst:
    sc = subcat[s]
    print("sub ID", sc.subcatchmentid)  # it prints all subcatchment IDs 
    for step in sim:
        print(sc.subcatchmentid, sim.current_time, sc.runoff)
        pass

**Console Output:** 
It only produced S1's runoff results at each time step, followed by sub ID list:
S1 2022-02-11 05:59:56 0.0
S1 2022-02-11 05:59:57 0.0
S1 2022-02-11 05:59:58 0.0
S1 2022-02-11 05:59:59 0.0
sub ID S2
sub ID S3
sub ID S4
sub ID S5
sub ID S6
...
Process finished with exit code 0

**Desired output**
runoff rate at each time step for every subcatchment 

**Quesitons:**
1. Is it possible to instantiate all subcatchments through a loop in "for step by sim"? 
2. Is there a function for obtaining final cumulative runoff volume of each subcatchment without going through each time step? 
3. If simulation can only be run through sim.excute() or "for step in sim", is it possible to save the simulation result files (.rpt, .out) with a different name? 
3. Is it better to create new inp files with new LID parameter, use pyswmm sim.excute(), then use the output file processing function of pyswmm to read results of all subcatchments? 

**Desktop**
OS:Windows 11 Pro
Python Version [Python 3.8]

Apologies for the long question list. Any comments or advice would be much appreciated! Thank you!

Best Regards,
Vivian

**Comment by jennwuu** [comment, marked as answer]
Hello Vivian,

Thanks for using pyswmm. 

1. Yes, it is possible to instantiate all subcatchments for a loop. I think you are doing that in the example provided. Let me know if I misunderstood your question.

```
with Simulation('template.inp') as sim:
    subcatchments_list = Subcatchments(sim)
    subcatchment_objects = [sub for sub in subcatchments_list]
    for step in sim:
        for sc in subcatchment_objects:
            print(sc.subcatchmentid, sim.current_time, sc.runoff)
```

2. You can track each flow rate and sum up the data during simulation iteration. Or you can use [statistics](https://pyswmm.readthedocs.io/en/stable/reference/subcatchments.html#pyswmm.subcatchments.Subcatchment.statistics).

```
from pyswmm import Simulation, Subcatchments

with Simulation('template.inp') as sim:
    subcatchments_list = Subcatchments(sim)
    subcatchment_objects = [sub for sub in subcatchments_list]
    subcatchment_stats = dict()
    for step in sim:
        for sc in subcatchment_objects:
            subcatchment_stats[sc.subcatchmentid] = sc.statistics
        
    for sc_id in subcatchment_stats:
        print(sc_id, subcatchment_stats[sc_id])
```

3. Yes, you can save the file under different .out (parameter: outputfile) or .rpt (reportfile). When [instantiating Simulation](https://pyswmm.readthedocs.io/en/stable/reference/simulation.html#pyswmm.simulation.Simulation) object, you can specify reportfile (.rpt name) or outputfile (.out name) in the parameters.

```
from pyswmm import Simulation, Subcatchments

with Simulation(inputfile='template.inp', reportfile='sloth.rpt', outputfile='potato.out') as sim:
    for step in sim:
        pass
```

4. You can consider processing LID data directly during simulation since pyswmm allows you to access model data during simulation). You can track the results in .json or .csv as you iterate through different simulations. This will save you a step from processing the .out file.

**Reply by vww-wrc** [reply]
Hi Jen,

Thank you very much for your prompt reply and example codes! Really appreciate it.  

Re. Question 1, thank you for your code, it works like a charm.  My original code placed the simulation "for step in sim" under the loop of subcatchments and it only produced the result of S1, not the rest of the subcatchments. I see that the code you provided loops through all the subcatchments at each time step, which is great.  

I do have an additional question if you don't mind: I used sub.runoff and sub.statistics["runoff"] to obtain runoff rates at each time step for a subcatchment, but I am getting different runoff rates at the same time step from these two functions. There were even negative values for runoff rates at some time steps. The total sum of runoff rates from all time steps are also far off from what I obtained from SWMM interface. Are sub.runoff and sub.statistics["runoff"] supposed to produce different results? Or is there something wrong in my code or in my .inp file?  

**Code**
![code](https://user-images.githubusercontent.com/83201399/154051188-e5fc6b72-ef98-4f52-a6d8-398333c09478.png)

**Output:**
![negative runoff rate](https://user-images.githubusercontent.com/83201399/154051409-cd66a588-d200-46aa-ab2b-3872eb2a171c.png)


Thank you!

Best Regards,
Vivian

**Reply by jennwuu** [reply]
Hello Vivian,

No problem. Happy to help with any questions.

The example I provided for statistics has a mistake. The statistics is only allowed at the end of simulation.
It is strange that the method doesn't throw an error when we're trying to access it during simulation. I will look into that.

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/toolkit.c#L2486-L2488

```
from pyswmm import Simulation, Subcatchments

with Simulation('template.inp') as sim:
    subcatchments_list = Subcatchments(sim)
    subcatchment_objects = [sub for sub in subcatchments_list]

    for step in sim:
        pass
        
    for sc in subcatchment_objects:
        print(sc.statistics)
```

**Reply by vww-wrc** [reply]
Hello Jenn,

Thank you very much! So the statistics["runoff"] at the end of simulation is the cumulative runoff volume?  The runoff results I got from statistics value seems to be three decimal places smaller than those obtained from using other approaches (see table below). Is the unit for this supposed to be 10^3 Litres or cubic feet?  
I also used sub.runoff to calculate total runoff volume and peak flow rate (see codes below), but the results from this approach seem to be a bit different from the results from using the statistics and outputfile... Should I stick to using the result from statistics? 
 
Sorry for raising so many questions. Thank you so much for your patience and support!!!

Best,
Vivian 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wenhu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wenhu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Functions for runoff in Subcatchment S55 | Peak rate (LPS) | Runoff volume   (Litre)
-- | -- | --
Pyswmm   sub_object.runoff | 332.26 | 534064.64
Pyswmm   sub_object.statistics | 334.23 | 568.71
Pyswmm   outputfile SubcatchAttribute.RUNOFF_RATE | 334.23 | 568599.99
SWMM5 | 334.23 | 568601.71



</body>

</html>

![code](https://user-images.githubusercontent.com/83201399/154187620-ae9b5af8-22a8-4d92-80ff-8f34588634e5.png)

**Reply by jennwuu** [reply]
Hey Vivian, 

This is really helpful for pyswmm as well. It's good to identify any potential bugs for the next user. 

If I understood the table correctly, the peak rate (lps) are consistent for all methods, but the runoff volume differ between the methods.

SWMM5 value should be the correct value. The runoff volume calculated  from the outfile method matches SWMM5 value (the difference is due to routing timestep versus reporting timestep).

The statistics method for runoff volume should be cumulative volume. 
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/stats.c#L996-L997

The internal engine units for runoff volume are cubic feet (ft3). 
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/stats.c#L352

The API returns the data by multiplying UCF(VOLUME). 
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/swmm5.c#L131

It looks like your units are SI, so UCF(VOLUME) multiplier is 0.02832. 
- ft 3 = 28.3168 L 
- ft 3 / 1000 = 0.02832 L 

It looks like the units are L/1000. I think you might see this unit in the .rpt file.  

I am surprised that runoff calculated by stepping through the simulation is significantly different compared to the outputfile vs. SWMM5 method. It would be interesting to graph the runoff rate extracted from iterating through simulation vs. the runoff rate extracted from the output method. Is it possible for you to share the template.inp file? I want to look into that a bit more detail.

**Reply by jennwuu** [reply]
Oh you probably need to multiple runoff rate by the timestep to get the volume for better comparison. Let me know how that value looks after you convert runoff rate to runoff volume.

Take a look at sim.current_time and the reporting timestep you selected in the SWMM model. When you iterate through Simulation using pyswmm, you will get results for every routing time step. 

In the outfile, you will get results for every reporting time step. 


If you want to change pyswmm to iterate through the simulation through a larger time step, check out [step_advance](https://pyswmm.readthedocs.io/en/stable/reference/simulation.html#pyswmm.simulation.Simulation.step_advance) method. 

For example, if you use sim.step_advance(300), you can iterate through your simulation every 5 minute instead of routing time step.

**Comment by vww-wrc** [comment]
Hi Jenn,

Thanks for your prompt response! I tried to upload my inp file to Github but the file type is not supported. Please see the attached template inp file.

Yes, the sub_object.runoff function produced different results to others, even for peak rate (minor difference). In terms of reporting time step and routing time step, I had specifically set them both to 1 second (00:00:01) in my template.inp to avoid causing difference in the results, so the values you see from the table were all calculated based on runoff rate a time step of 1sec.  And since the time step is 1 sec, the runoff volume value should be the same as the sum of runoff rate value at LPS because x times 1 still equals x. So I was really confused about the difference in the runoff volume between different approaches... Would be great to understand what caused the difference.

Re. unit, it's good to know that statistics runoff volume is in ft3. But I just multiplied the cumulative runoff volume from statistics 568.71 ft3 with 28.3168, it gives 16104.05L, which is still quite different to the SWMM5 results (568601.71L). With the SWMM model's unit set to LPS, the rpt file usually provides total runoff volume in 10^6 ltr (1,000,000), so it's giving 0.57 in my rpt file for Subcatchment S55 (see attached rpt file).

Thank you!

Best,
Vivian
________________________________

**Reply by jennwuu** [reply]
Hey Vivian, 

I didn't see the model .inp and .rpt. Can you share the model over email if possible? I can be reached at - removed -. Thanks!

**Reply by jennwuu** [reply]
I meant to say that the units for 568.71 is in 1000 L, so you need to multiple by 1000 to get the units in litres. The volume by pyswmm statistics method should be 568710 L. This volume is similar to the SWMM5 value: 568601.71 L. For your SWMM5 result, what SWMM5 version are you using? 

In the swmm .rpt file, you can see Subcatchment S55 total runoff is 0.57 10^6 L. 

```
  ***************************
  Subcatchment Runoff Summary
  ***************************
  
  ------------------------------------------------------------------------------------------------------------------------------
                            Total      Total      Total      Total     Imperv       Perv      Total       Total     Peak  Runoff
                           Precip      Runon       Evap      Infil     Runoff     Runoff     Runoff      Runoff   Runoff   Coeff
  Subcatchment                 mm         mm         mm         mm         mm         mm         mm    10^6 ltr      LPS
  ------------------------------------------------------------------------------------------------------------------------------
  S55                       91.10       0.00       0.00      10.41      86.49       0.00      56.73        0.57   366.88   0.623
```

**Reply by jennwuu** [reply]
In the outputfile code, you should update datetime(2022, 2, 11, 0, 1) to datetime(2022, 2, 11, 0, 0, 1) to represent 1 second instead of 1 minute.

**Reply by vww-wrc** [reply]
Thanks for the clarification on units, Jenn - super helpful!

Re. outputfile datetime, well spotted! I will fix that. In my model there is no rainfall in the first four minutes so this should not have impacted the runoff results calculation but always best to include the full time series. Thank you!

P.S. My SWMM5 version is 5.1.12. 

Best,
Vivian

**Reply by jennwuu** [reply]
I need to look into runoff method for subcatch some more, but I think there is a bug in the runoff method. I think the missing runoff rate (and runoff volume) when using `subcatch.runoff` is coming from the LID drain layer. I see some LID usage on subcatchment S55 in the file you shared with me.

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/output.c#L588

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/subcatch.c#L866-L932

https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/subcatch.c#L894-L905

Thank you for catching this! I will need to create an issue in OWA/SWMM and investigate more. In the meantime, I think outfile or subcatchment statistics is the best method to use for now.

CC: @bemcdonnell

**Reply by vww-wrc** [reply]
Thanks Jenn. In the template.inp I sent to you, all subcatchments have a LID in them, except for one (SubID:"NoLID"). I just did a quick run on this subcatchment's runoff volume and peak flow rates - see table below: subcatch.statistics is now producing the same result for peak rate as subcatch.runoff. But  the runoff volume from both subcatch.runoff and subcatch.statistics are different to the results from output file and SWMM5. 

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/wenhu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/wenhu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Subcatchment   NoLID | Peak rate (LPS) | Runoff volume   (Litre)
-- | -- | --
Pyswmm   sub_object.runoff | 423.2740089 | 911454.9409
Pyswmm   sub_object.statistics | 423.2740089 | 911551.5266
Pyswmm   outputfile SubcatchAttribute.RUNOFF_RATE | 423.2740173 | 911213.1884
SWMM5 version 5.1.12. | 423.2740173 | 911213.1884



</body>

</html>

**Reply by dickinsonre** [reply]
@vww-wrc they look like the same numbers to me - I think you are comparing too many decimal places.  @jennwuu https://medium.com/a-microbiome-scientist-at-large/why-dont-birds-or-sloths-fall-out-of-trees-10f94357ea4d

**Reply by jennwuu** [reply]
@dickinsonre That is so cool!! Thanks for sharing the article

**Reply by jennwuu** [reply]
@vww-wrc Great test results! This is super helpful because it helps me isolate the exact difference. The first set was showing differences in runoff rate / volume methods due to LID implementation. 

This set shows the difference even without LID implementation. 

The difference you are seeing between `runoff` method and `outputfile` method is due to these line:
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/025337c026dc2b6d493fb23ab05016329f7547cb/src/solver/subcatch.c#L903-L905

When runoff rate is lower than MIN_FLOW, the runoff rate is set to 0. You will see this difference if you compare the runoff rate vs the outputfile rate. The min_flow rate is 0.002492 ft3/s (1 ha = 107639 ft2, MIN_FLOW is 2.31481E-08, runoff > (2.31481E-08 * 107639 = 0.002492 ft3/s). 

If you set your runoff minimum rate to 0.002492 ft3/s, the results are match closer. Although the results were pretty similar.

```
with Simulation('Template.inp') as sim:
    no_lid = Subcatchments(sim)['NoLID']
    min_flow = 2.31481*10**(-8)
    no_lid_min_rate = min_flow * no_lid.area * 107639 # cfs
    no_lid_min_rate = no_lid_min_rate * 28.316847 # cfs to lps
    for step in sim:
        timestamps.append(sim.current_time)

        runoff = no_lid.runoff
        if runoff < no_lid_min_rate:
            data.append(0)
        else:
            data.append(no_lid.runoff)

    print(sum(data))
```

Looking into statistics method now...

I don't think we should cap the runoff rate due to this limitation. Thoughts @bemcdonnell?

**Reply by jennwuu** [reply]
From the .rpt file, the statistics generated for NoLID subcatchment is 0.91 10^6 litre, but unfortunately the decimal places do not show the exact value.

```
    ------------------------------------------------------------------------------------------------------------------------------
                            Total      Total      Total      Total     Imperv       Perv      Total       Total     Peak  Runoff
                           Precip      Runon       Evap      Infil     Runoff     Runoff     Runoff      Runoff   Runoff   Coeff
  Subcatchment                 mm         mm         mm         mm         mm         mm         mm    10^6 ltr      LPS
  ------------------------------------------------------------------------------------------------------------------------------
  NoLID                     91.10       0.00       0.00       0.00      91.14       0.00      91.14        0.91   423.27   1.000
  
```

In SWMM5 engine, subcatchment runoff volume statistics are calculated by summing runoff and lid drain together:
https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/blob/develop/src/solver/subcatch.c#L745-L749

Unlike the output code, there is no min flow rate that would set very small flow rate to zero. So the differences you are seeing between `statistics` and `outputfile` method is probably explained by the flow rate limitation that we saw in `runoff` method. This also helps to explain why the volumes are closer between `runoff` and `statistics` methods. Overall, I think these values are similar when there are no LIDs on the subcatchment. But we need to update runoff rate to include lid drain rates.

**Reply by vww-wrc** [reply]
Thank you very much Jenn! It's great to understand where the differences in the results (though very minimum) are coming from. I will stick to using the statistics and outputfile processing approaches for now and look forward to seeing the update in runoff rate in the future. Thumbs up for your great work and I really appreciate your help on this! 

Best,
Vivian

**Reply by vww-wrc** [reply]
Hi Jenn, sorry I have one more question regarding runoff rates... I am tracking runoff rate over time and I am getting negative values again in runoff reported in pyswmm. The subcatchment HC600_1Clay has a bioretention (overflow and underdrain all flows to the same outlet as the subcatchment). I have run my model in SWMM 5.1.015 and the runoff should be positive. The values also look a bit far off (e.g. at 07:00:00 SWMM reported 16.66LPS but pySWMM reported -3.35 e-16). Would appreciate advice on why this is happening and how to fix it. Thank you! 

Code:
inpmodel = "template.inp"

with Simulation(inpmodel) as sim:
    sub = Subcatchments(sim)["HC600_1Clay"]
    for step in sim:
        print(sim.current_time, sub.runoff)

Results:  at the beginning of the simulation:
![image](https://user-images.githubusercontent.com/83201399/163503805-2fe61542-9e28-4079-b2da-e26dd30173a5.png)

Runoff time series from SWMM 5.1.015
![image](https://user-images.githubusercontent.com/83201399/163503407-5443c2d1-41c5-43ac-958e-732788b740b1.png)

![image](https://user-images.githubusercontent.com/83201399/163503065-eab410ff-e727-4e4d-860b-46f0d0745ba7.png)

**Reply by jennwuu** [reply]
Hey @vww-wrc,

No worries. I think you are seeing the issue mentioned earlier where the lid runoff is not included in the runoff rate. I assume the difference increases when the LID covers the entire subcatchment (e.g. zero or close to zero runoff) for subcatchment. 

I will create an issue for this in OWA-SWMM: https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/issues/370

**Comment by dickinsonre** [comment]
I will just add that it common to have time steps with no rainfall using real storms.  The time steps with no rainfall can happen at any time.
